### PR TITLE
Add clang tidy

### PR DESCRIPTION
**Pull request by Alex-PLACET**
Implement #83 
Point to discuss:
[ ] Are you ok to deactivate some cheks as the magic numbers for the tests ?

### CODE REVIEW

File: .github/workflows/analysis.yaml [hunk 1]
```diff
@@ -27,7 +28,7 @@ jobs:
         cache-downloads: true
 
     - name: Configure using CMake
-      run: cmake -G Ninja -Bbuild -DCMAKE_BUILD_TYPE:STRING=Debug -DCMAKE_INSTALL_PREFIX=$CONDA_PREFIX -DBUILD_TESTS=ON
+      run: cmake -G Ninja -Bbuild -DCMAKE_BUILD_TYPE:STRING=Debug -DCMAKE_INSTALL_PREFIX=$CONDA_PREFIX -DBUILD_TESTS=ON -DCMAKE_EXPORT_COMPILE_COMMANDS=ON
```
Comment (jjerphan) [inline .github/workflows/analysis.yaml:31]: Should we turn the linter ON in another PR?

File: cmake/clang-tidy.cmake [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+find_program(CLANG_TIDY clang-tidy
+             VALIDATOR check_clang-tidy_version)
+
+if(NOT CLANG_TIDY)
+    message(WARNING "❗clang-tidy with version >= ${CLANG-TIDY_MINIMUM_MAJOR_VERSION} not found")
+
+    print_clang_tidy_install_instructions()
+else()
+    get_clang_tidy_version(${CLANG_TIDY})
+    message(STATUS "✅ clang-tidy found at ${CLANG_TIDY} | version: ${CLANG_TIDY_MAJOR_VERSION}.${CLANG_TIDY_MINOR_VERSION}.${CLANG_TIDY_PATCH_VERSION}")
```
Comment (jjerphan) [inline cmake/clang-tidy.cmake:61]: For consistency with what we have for `clang-format`:
```suggestion
    message(STATUS "✅ clang-tidy (version: ${CLANG_TIDY_MAJOR_VERSION}.${CLANG_TIDY_MINOR_VERSION}.${CLANG_TIDY_PATCH_VERSION}) found at ${CLANG_TIDY}")
```

File: cmake/clang-tidy.cmake [hunk 3]
```diff
@@ -0,0 +1,117 @@
+if(CMAKE_GENERATOR MATCHES "Ninja|Unix Makefiles")
+    message(STATUS "🔧 Compile command will be used to enable clang-tidy")
+    set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
+else()
+    message(WARNING "🚧 Compile command can't be used because the CMAKE_GENERATOR is ${CMAKE_GENERATOR}.
+You have to use Ninja or Unix Makefiles.
+Without compile command, clang-tidy will not work.")
+endif()
```
Comment (jjerphan) [inline cmake/clang-tidy.cmake:8]: Can those messages be explicit and explain what the "Compile command" here is?

File: cmake/clang-tidy.cmake [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    set(CLANG_TIDY_MAJOR_VERSION ${CMAKE_MATCH_1} PARENT_SCOPE)
+    set(CLANG_TIDY_MINOR_VERSION ${CMAKE_MATCH_2} PARENT_SCOPE)
+    set(CLANG_TIDY_PATCH_VERSION ${CMAKE_MATCH_3} PARENT_SCOPE)
+endfunction()
+
+function(check_clang-tidy_version validator_result_var item)
+    set(${validator_result_var} FALSE PARENT_SCOPE)
+    get_clang_tidy_version(${item})
+    if (CLANG_TIDY_MAJOR_VERSION LESS CLANG-TIDY_MINIMUM_MAJOR_VERSION)
+        message(DEBUG "clang-tidy found at ${item} | version: ${CLANG_TIDY_MAJOR_VERSION}.${CLANG_TIDY_MINOR_VERSION}.${CLANG_TIDY_PATCH_VERSION}")
```
Comment (jjerphan) [inline cmake/clang-tidy.cmake:28]:
```suggestion
        message(DEBUG "clang-tidy (version: ${CLANG_TIDY_MAJOR_VERSION}.${CLANG_TIDY_MINOR_VERSION}.${CLANG_TIDY_PATCH_VERSION}) found at ${item}")
```

File: cmake/clang-tidy.cmake [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+        message(DEBUG "but version is lower than ${CLANG-TIDY_MINIMUM_MAJOR_VERSION}")
```
Comment (jjerphan) [inline cmake/clang-tidy.cmake:29]:
```suggestion
        message(DEBUG "but clang-tidy with version >= ${CLANG-TIDY_MINIMUM_MAJOR_VERSION} must be used.")
```

**Review by jjerphan — APPROVED** [review]
LGTM modulo a few comments and suggestions.

Thank you, @Alex-PLACET.

File: .github/workflows/analysis.yaml (hunk 1 above)
Comment (Alex-PLACET) [reply .github/workflows/analysis.yaml:31]: This cmake option is for the developer, we don't need it in the CI as the CI don't run the linting cmake targets but clang-tidy directly

File: cmake/clang-tidy.cmake (hunk 3 above)
Comment (Alex-PLACET) [reply cmake/clang-tidy.cmake]: Ok 👍

**Review by Klaim — APPROVED** [review]
LGTM

> Are you ok to deactivate some cheks as the magic numbers for the tests ?

I dont understand the question?

**Comment by Alex-PLACET** [comment]
@Klaim it's about using or not magic number: number not associated with a variable

**Pull request merged**